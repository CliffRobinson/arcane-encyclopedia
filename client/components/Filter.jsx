import React from "react";

class Filter extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            function:"customTextFilter",
            key:"type_line",
            value:"",
            exact:false,
        };
        this.handleSelectChange = this.handleSelectChange.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
    }

    handleSelectChange(e, props){
        //console.log("Adding", e.target.value, "to filter", props.i);
        const funcKey = e.target.value.split("--");
        const obj = {
            function:funcKey[0],
            key:funcKey[1],
            value:this.state.value,
            exact:(false || funcKey[2])
        };
        this.setState(obj);
        props.addFilter(obj, props.i);
    }
    
    handleInputChange(e, props){
        //console.log("Adding", e.target.value, "to filter", props.i);
        const value = this.state.function.includes("number") ? parseFloat(e.target.value) : e.target.value;
        const objNixilis = {
            function:this.state.function,
            key:this.state.key,
            value:value,
            exact:this.state.exact
        };
        this.setState(objNixilis);
        props.addFilter(objNixilis, props.i);
    }

    render() {
        return <span className={"filter"}>
            <select className={`select ${this.props.size}`} onChange={(e) => this.handleSelectChange(e, this.props)}>
                <option value="customTextFilter--type_line"> Type Line contains </option>
                <option value="customTextFilter--name"> Name contains</option>
                <option value="customTextFilter--oracle_text"> Oracle Text Contains</option>
                <option value="customTextFilter--rarity--true"> Rarity is</option>
                <option value="numberLessFilter--cmc"> CMC less than </option>
                <option value="numberEqualsFilter--cmc"> CMC equals </option>
                <option value="numberMoreFilter--cmc"> CMC more than </option>
                <option value="numberLessFilter--usd"> Price less than </option>
                <option value="numberMoreFilter--usd"> Price more than </option>
            </select>
            <input className={`${this.props.size}`} type="text" onChange={(e) => this.handleInputChange(e, this.props)}>
    
            </input>
        </span>;
    }
}

export default Filter;