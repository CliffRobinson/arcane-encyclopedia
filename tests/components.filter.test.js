import React from 'react';
import {shallow, mount} from "enzyme";

import Filter from "../client/components/filter";

test("filter matches snapshot", ()=> {
    expect(
        toJson(shallow(<Filter />))
    ).toMatchSnapshot();
});

test("handleSelectChange behaves correctly", ()=> {
    //Arrange
    const fakeEvent = {
        target: {
            value: "customTextFilter--rarity--true"
        }
    };
    const fakeAddFilter = jest.fn();
    const props = {
        addFilter:fakeAddFilter,
        i:1
    };
    const testFilter = mount(<Filter />);
    const expectedState = {
        function:"customTextFilter",
        key: "rarity",
        value:"",
        exact: "true"
    };
    //Act
    testFilter.instance().handleSelectChange(fakeEvent, props);
    //Assert
    expect(testFilter.instance().state).toEqual(expectedState);
    expect(fakeAddFilter).toBeCalledWith(expectedState, 1);
});

test("handleInputChange behaves correctly for text", ()=> {
    //Arrange
    const fakeEvent = {
        target: {
            value: "human"
        }
    };
    // const fakeDispatch = jest.fn();
    const fakeAddFilter = jest.fn();
    const props = {
        // dispatch: fakeDispatch,
        addFilter:fakeAddFilter,
        i:1
    };
    const testFilter = mount(<Filter />);
    const expectedState = {
        function:"customTextFilter",
        key: "type_line",
        value:"human",
        exact: false
    };
    //Act
    testFilter.instance().handleInputChange(fakeEvent, props);
    //Assert
    expect(testFilter.instance().state).toEqual(expectedState);
    expect(fakeAddFilter).toBeCalledWith(expectedState, 1);
});

test("handleInputChange behaves correctly for numerics", ()=> {
    //Arrange
    const fakeEvent = {
        target: {
            value: "2"
        }
    };
    // const fakeDispatch = jest.fn();
    const fakeAddFilter = jest.fn();
    const props = {
        // dispatch: fakeDispatch,
        addFilter:fakeAddFilter,
        i:1
    };
    const testFilter = mount(<Filter />);
    const expectedState = {
        function:"customNumberFilter",
        key: "usd",
        value: 2.0,
        exact: false
    };
    //Act
    testFilter.instance().setState({
        function:"customNumberFilter",
        key: "usd"
    });
    testFilter.instance().handleInputChange(fakeEvent, props);
    //Assert
    expect(testFilter.instance().state).toEqual(expectedState);
    expect(fakeAddFilter).toBeCalledWith(expectedState, 1);
});

test("changing the select calls handleSelectChange", ()=> {
    //Arrange
    const fakeEvent = {target:{}};
    const testFilter = shallow(<Filter />);
    const spy = jest.spyOn(testFilter.instance(), "handleSelectChange").mockImplementation(() => true);
    //Act
    testFilter.find("select").at(0).props().onChange(fakeEvent);
    //Assert
    expect(spy.mock.calls[0][0]).toBe(fakeEvent);
});

test("changing the input calls handleSelectChange", ()=> {
    //Arrange
    const fakeEvent = {target:{}};
    const testFilter = shallow(<Filter />);
    const spy = jest.spyOn(testFilter.instance(), "handleInputChange").mockImplementation(() => true);
    //Act
    testFilter.find("input").at(0).props().onChange(fakeEvent);
    //Assert
    expect(spy.mock.calls[0][0]).toBe(fakeEvent);
});