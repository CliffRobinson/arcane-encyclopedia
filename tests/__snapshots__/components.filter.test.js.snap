// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter matches snapshot 1`] = `
<span
  className="filter"
>
  <select
    className="select undefined"
    onChange={[Function]}
  >
    <option
      value="customTextFilter--type_line"
    >
       Type Line contains 
    </option>
    <option
      value="customTextFilter--name"
    >
       Name contains
    </option>
    <option
      value="customTextFilter--oracle_text"
    >
       Oracle Text Contains
    </option>
    <option
      value="customTextFilter--rarity--true"
    >
       Rarity is
    </option>
    <option
      value="numberLessFilter--cmc"
    >
       CMC less than 
    </option>
    <option
      value="numberEqualsFilter--cmc"
    >
       CMC equals 
    </option>
    <option
      value="numberMoreFilter--cmc"
    >
       CMC more than 
    </option>
    <option
      value="numberLessFilter--usd"
    >
       Price less than 
    </option>
    <option
      value="numberMoreFilter--usd"
    >
       Price more than 
    </option>
  </select>
  <input
    className="undefined"
    onChange={[Function]}
    type="text"
  />
</span>
`;
